Feature/issue 5 current article exposed as json

gitumarkk opened the pull request:
Fix #5 
- Created articles API in article/views.py
- Created unittests in a test_views.py
- Added url resource to url.py

gitumarkk committed:
Created Unittests for API (TDD)

gitumarkk committed:
Added URL to find required API function in views

gitumarkk committed:
Created the API fn in views called get_articles()

imsickofmaps commented on articles/urls.py:
```diff
@@ -0,0 +1,5 @@
+from django.conf.urls import patterns, url
+
+urlpatterns = patterns('',
+    url(r'^articles/$', 'articles.views.get_articles'),
```
drop the /  and make all references singular as we're always working with one article

gitumarkk committed:
Change way header is checked

gitumarkk committed:
Changed the urls to r'^articles from r'^articles/

gitumarkk committed:
Changed unittests to account for urls

gitumarkk committed:
Made articles to article (singular)

imsickofmaps commented:
@smn ready for your eyes - made my comments

smn commented on articles/tests/test_views.py:
```diff
@@ -0,0 +1,73 @@
+        Article.objects.create(article="Test5",
+                               publish=False,
+                               publish_at=
+                               (timezone.now() - datetime.timedelta(days=2)))
```
fixtures will probably make this easier and more reusable?

smn commented on articles/tests/test_views.py:
```diff
@@ -0,0 +1,73 @@
+
+        response = self.client.get("/api/article")
+        self.assertEqual("application/json", response["Content-Type"])
+        self.assertIn('Test2', response.content)
```
You're getting back JSON, a better test is to do `json.loads(response.content)` which'll give you back a dictionary and do asserts on that?

gitumarkk replied:
Ok

smn commented on articles/tests/test_views.py:
```diff
@@ -0,0 +1,73 @@
+                               publish_at=
+                               (timezone.now() - datetime.timedelta(days=2)))
+
+        response = self.client.get("/api/article")
```
Please use Django's `reverse()` url helpers to avoid hard coding URLs anywhere.

gitumarkk replied:
Ok

smn commented on articles/tests/test_views.py:
```diff
@@ -0,0 +1,73 @@
+                               (timezone.now() + datetime.timedelta(seconds=60)))
+
+        response = self.client.get("/api/article")
+        self.assertIn("application/json", response.items()[0])
```
check for `response['Content-Type']`, the order of headers in an HTTP response isn't guaranteed (or important).

gitumarkk replied:
A later commit [https://github.com/praekelt/go-nike-ghr/commit/b909e962d6593b838fde5e603c82f216c0ba1a27#L0R38] sorts this issue out.

imsickofmaps replied:
@gitumarkk - no it doesn't, otherwise @smn wouldn't see it in the diff! This is a second reference to the same issue. :smile:

smn commented on articles/tests/test_views.py:
```diff
@@ -0,0 +1,73 @@
+
+        response = self.client.get("/api/article")
+        self.assertIn("application/json", response.items()[0])
+        self.assertIn("Sorry there's no article this week", response.content)
```
test for json

gitumarkk replied:
Ok

smn commented on articles/urls.py:
```diff
@@ -0,0 +1,5 @@
+from django.conf.urls import patterns, url
+
+urlpatterns = patterns('',
+    url(r'^article$', 'articles.views.get_article'),
```
I know @imsickofmaps advised to remove the trailing slash, I'm suggesting you add it back in. Django by default ships with a middleware that [appends slashes](https://docs.djangoproject.com/en/dev/ref/settings/#append-slash) if a page 404s to see if it does exist with a trailing slash. This is a Django convention we follow and has to do with keeping the paths of relative paths more predictable.

gitumarkk replied:
Cool, will revert back

smn commented on articles/views.py:
```diff
@@ -1 +1,22 @@
+                    filter(publish=True).
+                    filter(publish_at__lte=timezone.now()).
+                    filter(publish_at__gte=timedelta).
+                    order_by('publish_at')[0])
```
You want to use [`.exists()`](https://docs.djangoproject.com/en/1.5/ref/models/querysets/#django.db.models.query.QuerySet.exists) here, it is more efficient and prevents the need for the IndexError.

gitumarkk replied:
Will change it.

smn commented on articles/views.py:
```diff
@@ -1 +1,22 @@
-# Create your views here.
+from django.http import HttpResponse
+from articles.models import Article
+from django.utils import simplejson
```
We're using Python 2.6, it ships with the `json` module, rather use that than `simplejson` as the latter sometimes returns bytecode strings and sometimes unicode. It does this because of performance reasons, we'd rather take the performance hit for predictability.

smn commented on articles/views.py:
```diff
@@ -1 +1,22 @@
+        output = {"article":
+                  "Sorry there's no article this week, dial back soon!"}
+
+    return HttpResponse(simplejson.dumps(output),
```
Before you guys invent your own rest-ful API stuff, have a look at https://github.com/toastdriven/django-tastypie.

imsickofmaps replied:
@gitumarkk please follow the advice here - I've discussed on IRC with @smn and he's right. As usual.

gitumarkk replied:
@imsickofmaps shall I use django-tastypie or the django-rest-framework as was in the Django Text book

imsickofmaps replied:
@gitumarkk use tastypie

gitumarkk committed:
using django's reverse() instead of hard coding

gitumarkk committed:
Fixing application/json assert

gitumarkk committed:
Testing for JSON dictionary

gitumarkk committed:
Adding trailing slash to r'^article$'

gitumarkk committed:
Changing from simplejson to json

gitumarkk committed:
Replacing IndexError with queryset.exists()

gitumarkk committed:
pep8 changes

gitumarkk committed:
Deleted debugging print statement

gitumarkk committed:
Installed tastypie

Added it to settings and requirements

gitumarkk committed:
Created api file to handle api responses

Deleted content of views as they are no longer necessary

gitumarkk committed:
Updated unittest to use new API

imsickofmaps commented on ghr/urls.py:
```diff
@@ -11,8 +11,10 @@
     url(r'^celery/', include('celery_app.urls')),
+
+    # API links
+    url(r'^api/', include('articles.urls')),
```
Please make sure we structure the URLs for the API in a way that means when the next application gets added this still works.

gitumarkk committed:
Creating fixtures to test articles API

gitumarkk committed:
Changing url path to be defined by the Api

Changed from being defined by the urls directory

gitumarkk committed:
Updated unittests to include Api url path

gitumarkk committed:
Testing if URL path is what it should be

As the `reverse()` function is used, need to have a way to test the
correct URL path.

smn commented:
@imsickofmaps @gitumarkk Is this ready for re-review?

imsickofmaps commented:
@smn I've asked for a some comments to be added to the code but if you have a moment to review without them then as soon as they're added I'll merge

gitumarkk committed:
Adding comments to code

gitumarkk commented:
@imsickofmaps added comments and made latest commit of the code.

gitumarkk committed:
Made pep8 changes

imsickofmaps commented:
@smn Ready for review. Hopefully last.

gitumarkk committed:
Changed model structure and did Schemamigration

gitumarkk committed:
Updated unittests for changes

Changed unittests to take individual article pages as opposed to one
article

gitumarkk committed:
Modified unicode to take in 1st page

gitumarkk committed:
Changed API to output required field formats

gitumarkk committed:
Creating a TextArea widget for the page_# field

imsickofmaps commented:
@smn really ready. :grin:

smn commented on articles/admin.py:
```diff
@@ -1,4 +1,21 @@
+        if db_field.name == 'page_3':
+            formfield.widget = forms.Textarea(attrs=formfield.widget.attrs)
+        if db_field.name == 'page_4':
+            formfield.widget = forms.Textarea(attrs=formfield.widget.attrs)
```
This seems like a long way around for either (ModelAdmin.formfield_overrides](https://docs.djangoproject.com/en/1.5/ref/contrib/admin/#django.contrib.admin.ModelAdmin.formfield_overrides) or specifying a custom (ModelAdmin.form)[https://docs.djangoproject.com/en/dev/ref/contrib/admin/#django.contrib.admin.ModelAdmin.form] that sets the `widget=forms.TextArea` on the form fields.

If anything else then reduce this to:

``` python
if db_field.name in ['page_1','page_2','page_3','page_4']:
    formfield.widget = forms.Textarea(attrs=formfield.widget.attrs)
```

smn commented on articles/api.py:
```diff
@@ -0,0 +1,52 @@
+                 filter(publish_at__gte=timedelta).
+                 order_by('publish_at'))
+
+        if query.exists():
```
why is this necessary? Is returning a `query` that doesn't have any results not the same as returning `Article.objects.none()` ?

smn commented on articles/api.py:
```diff
@@ -0,0 +1,52 @@
+            # Need to create an empty object otherwise tastypie complains
+            return Article.objects.none()
+
+    def alter_list_data_to_serialize(self, request, data_dict):
```
My gut feel is that it'd be easier to leave this an API that just returns whatever it has and leaves the JS application to make sense of it.

smn commented on articles/tests/test_views.py:
```diff
@@ -0,0 +1,69 @@
+
+    def test_correct_url(self):
+        # Tests teh URL path to see if it is what its expected
+        response = self.client.get("/api/article/")
```
use `reverse()` please

smn commented:
:+1: other than my small-ish comments.

smn commented:
my comment with the `alter_list_data_to_serialize()` hook is certainly up for debate.

gitumarkk committed:
Removed query.exists() and returned query object

gitumarkk committed:
Made changes to admin Textarea Widget

gitumarkk committed:
Removed absolute path to test, using reverse()

gitumarkk commented:
@imsickofmaps made commit
